Allow more general input to ripser when metric='precomputed', improve/refactor check_point_clouds and add tests

ulupo opened the pull request:
**Types of changes**
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

**Description**
The initial thrust for this PR was the need to lift unnecessary input restrictions in `gtda.externals.ripser` when `metric='precomputed'`. Namely, `pairwise_distances(X, metric='precomputed')` will perform input validation and reject any`X` containing negative entries. But the theory of Vietoris-Rips filtrations and the C++ implementation of `ripser` handle these cases perfectly well.

However, simply removing the call to `pairwise_distances` in the case `metric='precomputed'` would have removed useful checks for input shape (square). The point of view was adopted that since `gtda.externals.ripser` is not exposed to the user via documentation, and is only meant to be called by `VietorisRipsPersistence` instances, these checks should in fact be moved to the input validation steps in that class, i.e. to the calls to `check_point_clouds`.

Hence, `check_point_clouds` was refactored as follows:
- The parameter `distance_matrix` was renamed `distance_matrices` (plural, agreeing with `check_point_clouds`).
- When `distance_matrices` is set to `True`, the shapes of entries are checked to be square.
- Warnings have been added to guide the user to correctly setting the `distance_matrices` parameter.
- The default behaviour of sklearn's `check_array` has been modified so that `force_all_finite=False` does *not* mean accepting NaN input (only infinite input is accepted). Note that `ripser` handles infinities quite correctly when distance/adjacency matrices are passed.
- The docstrings have been improved.

Finally, comprehensive tests for `check_point_clouds` were created.

**Checklist**
- [x] I have read the [guidelines for contributing](https://giotto-ai.github.io/gtda-docs/latest/contributing/#guidelines).
- [x] My code follows the code style of this project. I used `flake8` to check my Python changes.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed. I used `pytest` to check this on Python tests.

Commit:
Remove boolean mask on infinite values

Commit:
Refactoring of check_point_clouds

- Rename distance_matrix kwarg to distance_matrices
- Add checks for square input when distance_matrices is set to True
- More comprehensive/friendly error messages and warnings

Commit:
Rename distance_matrix to distance_matrices in homology/simplicial.py

Made necessary by renaming performed in 9bab872352cbbfd046425163c589c94e9ad37628

Commit:
Increase generality of ripser function when metric == 'precomputed'

Sklearn's pairwise_distances is no longer called when metric='precomputed'.

This is because pairwise_distances checks for positivity and finite input in that case, but this is unnecessarily restrictive for ripser. 

pairwise_distances also checks for square input so this is no longer checked in the case metric='precomputed'.  But the ripser function is not meant for use in isolation. It is meant to be called by a VietorisRipsPersistence instance, and input validation can be offloaded to that class. In particular, 9bab872352cbbfd046425163c589c94e9ad37628 introduces checks for square input.

Commit:
Add tests for check_point_clouds

wreise commented on gtda/externals/python/ripser_interface.py:
```diff
@@ -175,7 +176,10 @@ def ripser(X, maxdim=1, thresh=np.inf, coeff=2, metric="euclidean",
         dm = dperm2all[:, idx_perm]
     else:
-        dm = pairwise_distances(X, metric=metric)
+        if metric == 'precomputed':
+            dm = X
+        else:
+            dm = pairwise_distances(X, metric=metric)
```
Since we do checks ourselves anyway, could we use `pairwise_distances(X, metric=metric,  force_all_finite=False)` instead of the `if`?

ulupo replied:
In this way, we would still error on input with negative entries, due to https://github.com/scikit-learn/scikit-learn/blob/95d4f0841d57e8b5f6b2a570312e9d832e69debc/sklearn/metrics/pairwise.py#L1721. One of the main points of this PR is to avoid this.

wreise commented on gtda/utils/validation.py:
```diff
@@ -204,45 +220,90 @@ def check_point_clouds(X, distance_matrix=False, **kwargs):
-        return check_array(X, allow_nd=True, **kwargs_)
+            if X.ndim == 2:
+                extra_2D = \
+                    "\nReshape your input X using X.reshape(1, *X.shape) or " \
+                    "X[None, :, :] if X is a single point cloud/distance " \
+                    "matrix/adjacency matrix of a weighted graph."
```
Nice

wreise commented on gtda/utils/validation.py:
```diff
@@ -6,7 +6,9 @@
 import numpy as np
+
 from sklearn.utils.validation import check_array
+from sklearn.exceptions import DataDimensionalityWarning
```
Amazing!

wreise commented on gtda/utils/validation.py:
```diff
@@ -204,45 +220,90 @@ def check_point_clouds(X, distance_matrix=False, **kwargs):
+        for i, x in enumerate(X):
+            try:
+                xnew = _check_array_mod(x, **kwargs_, ensure_2d=True)
+                if distance_matrices:
+                    if not x.shape[0] == x.shape[1]:
+                        raise ValueError(
+                            f"All arrays must be square: {x.shape[0]} rows "
+                            f"and {x.shape[1]} columns found in this array.")
```
What do you think about including this in `_check_array_mod`, since `is_square` is a property of an individual array?

ulupo replied:
I would prefer it here as I didn't intend for `_check_array_mod` to add new parameters. Also: 1) the error message would have to be changed; 2) this would not cover the 3D case.

wreise commented on gtda/utils/validation.py:
```diff
@@ -204,45 +220,90 @@ def check_point_clouds(X, distance_matrix=False, **kwargs):
+            if not reduce(and_, (x.shape[1] == ref_dim for x in X[1:]), True):
+                warn(
+                    "Not all point clouds have the same embedding dimension.",
+                    DataDimensionalityWarning, stacklevel=2)
```
Could you detail the choice of `stacklevel` or point me to a reference, please? I am completely unfamiliar with that and could not find anything.

ulupo replied:
I basically stole this from sklearn's `check_array`, see e.g. https://github.com/scikit-learn/scikit-learn/blob/95d4f0841d57e8b5f6b2a570312e9d832e69debc/sklearn/utils/validation.py#L450.

wreise commented on gtda/utils/tests/test_validation.py:
```diff
@@ -54,3 +56,206 @@ def test_inputs_arrayStruc_V():
+        self.X_list_rectang_diff_cols = \
+            self.X_list_rectang[:-1] + [self.X_list_rectang[-1][:, :-1]]
+
+        N_extra = n_samples_extra * n_1_extra * n_2_extra
+        X_extra = np.arange(N_extra, dtype=float). \
+            reshape(n_samples_extra, n_1_extra, n_2_extra)
+        X_list_extra = []
+        for i in range(n_samples_extra):
+            X_list_extra.append(X_extra[i].copy())
+        self.X_list_tot = self.X_list + X_list_extra
```
I don't fully understand that. Why do we need those?

ulupo replied:
I wanted to cover the case of a list of square arrays of varying dimension.

wreise replied:
Yes, my bad.

wreise commented on gtda/utils/tests/test_validation.py:
```diff
@@ -54,3 +56,206 @@ def test_inputs_arrayStruc_V():
+    # Check that we error when force_all_finite is True or False
+    # 1) Array input
+    with pytest.raises(ValueError):
+        check_point_clouds(
+            ex.X, distance_matrices=True, force_all_finite=force_all_finite)
+    with pytest.raises(ValueError):
+        check_point_clouds(ex.X_rectang, force_all_finite=force_all_finite)
+    # 2) List input
+    with pytest.raises(ValueError):
+        check_point_clouds(ex.X_list, distance_matrices=True,
+                           force_all_finite=force_all_finite)
+    with pytest.raises(ValueError):
+        check_point_clouds(
+            ex.X_list_rectang, force_all_finite=force_all_finite)
```
I don't understand why these pass: why will `check_point_clouds` throw an error, if `force_all_finite=False`?

ulupo replied:
Correct, because there are NaNs this time. The behaviour is different from sklearn's `check_array`, and is the reason for creating `_check_array_mod`. In `check_array`, `force_all_finite=False` means that both `np.nan` and `np.inf` is allowed, while `force_all_finite='allow-nan'` means that `np.nan` is allowed but `np.inf` is not. I thought a more useful behaviour for us when `force_all_finite=False` would be: `np.inf` is OK, but `np.nan` is not.

wreise replied:
Thank you for the detailed explanation, it's clear.

wreise reviewed (commented):
Very very nice, also the extensive testing!
Only questions, not 'requesting changes'.

wreise reviewed (approved):
LGTM

Commit:
Refresh docstrings for VietorisRipsPersistence

- Mention "adjacency matrices of weighted undirected graphs" as well as distance matrices
- Mention compatibility with filtration in docstrings for fit and transform

Commit:
Fix inline comments

Commit:
Merge branch 'master' into improve_ripser_interface

Commit:
Merge branch 'master' into improve_ripser_interface
